src/subdomains/staking/infrastructure/staking-defichain.service.ts

```diff
@@ -142,4 +122,53 @@ export class StakingDeFiChainService {
       throw e;
     }
   }
+
+  // --- WITHDRAWALS --- //
+  async getPossibleWithdrawals(withdrawals: Withdrawal[]): Promise<Withdrawal[]> {
+    const dfiBalance = await this.whaleClient.getUtxoBalance(Config.staking.liquidity.address);
+    const dusdBalance = await this.whaleClient.getTokenBalance(Config.yieldMachine.liquidity.address, 'DUSD');
+
+    return [
+      ...this.getPossibleWithdrawalsFor(dfiBalance, 'DFI', withdrawals),
+      ...this.getPossibleWithdrawalsFor(dusdBalance, 'DUSD', withdrawals),
+    ];
+  }
+
+  private getPossibleWithdrawalsFor(balance: BigNumber, assetName: string, withdrawals: Withdrawal[]): Withdrawal[] {
+    const sortedWithdrawals = withdrawals.filter((w) => w.asset.name === assetName).sort((a, b) => a.amount - b.amount);
+
+    const possibleWithdrawals = [];
+    let withdrawalSum = 0;
+    for (const withdrawal of sortedWithdrawals) {
+      withdrawalSum += withdrawal.amount;
+      if (balance.minus(1).lt(withdrawalSum)) break;
```

Comment:
1 is min withdrawal size?

Reply:
That's just a buffer, could probably be smaller 🤔

Reply:
buffer for what? I don't quite understand why it is there, what should it prevent and why?

Reply:
In the end it's just a fee buffer. If we have 100 DFI on the liquidity address, we won't be able to pay out 100 DFI.